cdc/sink/schema_registry.go

```diff
@@ -0,0 +1,259 @@
+	RegistryID int64  `json:"id"`
+	Schema     string `json:"schema"`
+}
+
+// NewAvroSchemaManager creates a new AvroSchemaManager
+func NewAvroSchemaManager(registryURL string) (*AvroSchemaManager, error) {
+	registryURL = strings.TrimRight(registryURL, "/")
+	// Test connectivity to the Schema Registry
+	resp, err := http.Get(registryURL)
+	if err != nil {
+		return nil, errors.Annotate(err, "Test connection to Schema Registry failed")
+	}
+	defer resp.Body.Close()
+
+	text, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return nil, errors.Annotate(err, "Reading response from Schema Registry failed")
+	}
+
+	if string(text[:]) != "{}" {
+		return nil, errors.New("Unexpected response from Schema Registry")
+	}
+
+	log.Info("Successfully tested connectivity to Schema Registry", zap.String("registryURL", registryURL))
+
+	return &AvroSchemaManager{
+		registryURL: registryURL,
+		cache:       make(map[string]*schemaCacheEntry, 1),
+	}, nil
+}
+
+var regexRemoveSpaces = regexp.MustCompile(`\s`)
+
+// Register the latest schema for a table to the Registry, by passing in a Codec
+func (m *AvroSchemaManager) Register(tableName model.TableName, codec *goavro.Codec) error {
+	// The Schema Registry expect the JSON to be without newline characters
+	reqBody := registerRequest{
+		Schema:     regexRemoveSpaces.ReplaceAllString(codec.Schema(), ""),
+		SchemaType: "AVRO",
+	}
+	payload, err := json.Marshal(&reqBody)
+	if err != nil {
+		return errors.Annotate(err, "Could not marshal request to the Registry")
+	}
+	uri := m.registryURL + "/subjects/" + url.QueryEscape(tableNameToSchemaSubject(tableName)) + "/versions"
+	log.Debug("Registering schema", zap.String("uri", uri), zap.ByteString("payload", payload))
+
+	resp, err := http.Post(uri, "application/vnd.schemaregistry.v1+json", bytes.NewReader(payload))
+	if err != nil {
+		log.Warn("Failed to register schema to the Registry",
+			zap.String("uri", uri),
+			zap.ByteString("payload", payload))
+		return errors.Annotate(err, "Failed to register schema to the Registry")
+	}
+
+	defer resp.Body.Close()
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return errors.Annotate(err, "Failed to read response from Registry")
+	}
+
+	if resp.StatusCode != 200 {
+		log.Warn("Failed to register schema to the Registry, HTTP error",
+			zap.Int("status", resp.StatusCode),
+			zap.String("uri", uri),
+			zap.ByteString("requestBody", payload),
+			zap.ByteString("responseBody", body))
+		return errors.New("Failed to register schema to the Registry, HTTP error")
+	}
+
+	var jsonResp registerResponse
+	err = json.Unmarshal(body, &jsonResp)
+
+	if err != nil {
+		return errors.Annotate(err, "Failed to parse result from Registry")
+	}
+
+	if jsonResp.ID == 0 {
+		return errors.New(fmt.Sprintf("Illegal schema ID returned from Registry %d", jsonResp.ID))
+	}
+
+	log.Info("Registered schema successfully",
+		zap.Int("id", jsonResp.ID),
+		zap.String("uri", uri),
+		zap.ByteString("body", body))
```

Comment:
ByteString is not readable it produces something like "\u00FF". What about "zap.String"?

Reply:
ByteString is readable when it contains only printable characters, and per my understanding, a JSON body should not contain any unprintable characters. If it does, then something suspicious is definitely happening, so some log output is necessary.